### PR TITLE
Add port ranges to machine description

### PR DESCRIPTION
**Pull request by achilleasa**
This PR is part of the work for allowing juju to open/close machine ports in an endpoint-aware manner. To this end, juju must be able to accurately describe the set of opened port ranges for a machine by linking each opened port range with the set of endpoints that the range is exposed to.

The `Machine` type already includes information about port ranges but this information is associated with subnet IDs instead of endpoints. As the type could not be augmented to include the required information, this PR introduces a few new interfaces (`MachinePortRanges`, `UnitPortRanges` and `UnitPortRange`) and flags the old Port types as deprecated.

The new types require a new version (V2) for the type that describes machines. The `Machine` type now only exposes the new interfaces for port ranges. To maintain backwards compatibility, the V1 parsing code will transparently convert the legacy port range values into the new format by exploiting the observation that pre 2.9 controllers never used the multi-subnet functionality of the opened ports documents but instead assumed that all ports are opened to **all** subnets (in other words, the legacy port range lists only include a single document with an **empty** `SubnetID`). Consequently, we can simply convert the old port ranges as if they are exposed for **all** endpoints (signified by an empty endpoint name).

Finally, as a drive-by fix, the PR changes the model validation code so that mutable maps are not passed by value across methods.

### CODE REVIEW

**Comment by achilleasa** [comment]
@jameinel I am not sure why it was initially modeled in this way. 

My understanding from reading the juju bits is that having the unit in there allows us to reserve a particular port for that unit only and prevent other units from claiming the same port (the current behavior is to error if a unit attempts to open the same/conflicting port range with other units).  Currently, juju opens ports across all subnets but with the upcoming open-for-endpoint changes we would definitely need the unit in there to prevent a rogue/buggy charm (e.g. a subordinate) from messing up with the ports of another unit (e.g. unit X opens the port in the `dmz` space but unit Y opens the same port across all spaces).

Other benefits of having the unit in there include:
1) being able to detect which ports should be closed when a unit gets removed from a machine.
2) being able to correctly re-generate the ingress rules when the space topology (e.g. subnets get moved to a different space) and/or bindings change (juju bind, juju upgrade-charm --bind). If we only had the subnets to work with it would be very hard (if not impossible) to achieve this.

Note also that storing subnets (as we currently do) is problematic for the 2nd case because you need to diff the space topology (and then check the bindings) to figure out if a port needs to be opened in a different subnet. My plan for fixing this is to store the endpoints (hence this PR) and lazily resolve the subnets in the firewall worker whenever the rules need to be re-generated.

**Comment by SimonRichardson** [comment]
> 2nd case because you need to diff the space topology (and then check the bindings) 

This is harder than I was expecting, I've tried doing this previously!

File: machine.go [hunk 1]
```diff
@@ -588,13 +561,107 @@ func importMachineV1(source map[string]interface{}) (*machine, error) {
...
 	}
 
 	return result, nil
 
 }
 
+func machineV2(valid map[string]interface{}) (*machine, error) {
+	mach, err := machineV1(valid)
+	if err != nil {
+		return nil, err
```
Comment (SimonRichardson) [inline machine.go:589]:
```suggestion
		return nil, errors.Trace(err)
```

File: machine.go [hunk 2]
```diff
@@ -588,13 +561,107 @@ func importMachineV1(source map[string]interface{}) (*machine, error) {
...
+		return mach, nil
+	}
+
+	// V1 legacy ports have been converted to open port ranges by the V1
+	// machine parser. V2 machines *only* include port ranges so if present
+	// we can simply parse them and inject them into the parsed machine
+	// instance.
+	machPortRanges, err := importMachinePortRanges(machPortRangesSource)
+	if err != nil {
+		return nil, err
```
Comment (SimonRichardson) [inline machine.go:603]:
```suggestion
		return nil, errors.Trace(err)
```

**Comment by achilleasa** [comment]
$$merge$$

**Pull request merged**